[![Open In Colab](https://colab.research.google.com/assets/colab-badge.svg)](https://colab.research.google.com/github/miguepoloc/toma-decisiones-mcda/blob/main/03_vikor_iot_palmor.ipynb)

# VIKOR, caso IoT/WSN Palmor

Mismo caso y mismos datos que TOPSIS. Elección de tecnología de comunicación (**LoRaWAN, GSM/GPRS, Sigfox, Zigbee**) para una red de sensores IoT/WSN de monitoreo agroclimático en Palmor, corregimiento de Ciénaga (Sierra Nevada de Santa Marta, Magdalena), zona con conectividad limitada verificada vía MinTIC. Los 4 criterios (Alcance de comunicación, Autonomía de batería, Infraestructura/cobertura comercial en Colombia, Madurez/viabilidad comercial del proveedor) salieron del panel de evidencia de la Sesión 1. Los valores técnicos de la matriz de decisión son reales, verificados vía WebSearch (datasheets SIMCom/DigiKey, The Things Network, Lauridsen et al. 2019 *Sensors*/MDPI, noticias de apagado de 2G en Colombia). Contenido completo en la Sesión 3 (TOPSIS/VIKOR) del curso.

In [1]:
!pip install -q pyDecision


[notice] A new release of pip is available: 26.1.2 -> 26.2.1
[notice] To update, run: pip install --upgrade pip


In [2]:
import numpy as np
from pyDecision.algorithm import vikor_method

tecnologias = ["LoRaWAN", "GSM/GPRS", "Sigfox", "Zigbee"]
dataset = np.array([
    [10,   8,   2, 5],
    [10.5, 0.5, 3, 2],
    [40,   2,   5, 2],
    [0.07, 1.5, 2, 4],
])
pesos = [0.16, 0.25, 0.488, 0.102]
tipo = ["max", "max", "max", "max"]

## Cálculo de S, R, Q

In [3]:
s, r, q, _ = vikor_method(dataset, pesos, tipo, graph=False)

def a_dict(arr):
    return {tecnologias[int(idx) - 1]: val for idx, val in arr}

s_d, r_d, q_d = a_dict(s), a_dict(r), a_dict(q)
print("Ranking final VIKOR (menor Q es mejor):")
for t, qv in sorted(q_d.items(), key=lambda x: x[1]):
    print(f"  {t:10s} S={s_d[t]:.3f} R={r_d[t]:.3f} Q={qv:.3f}")

a1: 3.0
a2: 2.0
a3: 1.0
a4: 4.0
Ranking final VIKOR (menor Q es mejor):
  Sigfox     S=0.302 R=0.200 Q=0.000
  GSM/GPRS   S=0.796 R=0.325 Q=0.631
  LoRaWAN    S=0.608 R=0.488 Q=0.757
  Zigbee     S=0.899 R=0.488 Q=1.000


**Resultado esperado** (coincide con lo publicado en la Sesión 3 del curso): 1º Sigfox (S=0.302, R=0.200, Q=0.000) · 2º GSM/GPRS (S=0.796, R=0.325, Q=0.631) · 3º LoRaWAN (S=0.608, R=0.488, Q=0.757) · 4º Zigbee (S=0.899, R=0.488, Q=1.000). A diferencia de AHP/TOPSIS, VIKOR invierte el 2º y 3er lugar (GSM/GPRS supera a LoRaWAN) — el hallazgo central del día en la Sesión 3.